### PR TITLE
Feature/38206 pupil register cache

### PR DESCRIPTION
**Pull request by GeorgeChatzigiannis**
- Introduced service responsible for setting the pupil register in cache for a school
- Drop cache on pupil add(single-multiple)/edit, group add/edit/remove

### CODE REVIEW

File: admin/services/pupil-register-caching.service.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+   * @return {Array}
+   */
+  setPupilRegisterCache: async function (schoolId, pupilRegisterViewData) {
+    if (!schoolId) {
+      throw new Error('School id not found in session')
+    }
+    if (!pupilRegisterViewData || pupilRegisterViewData.length === 0) {
+      throw new Error('Pupil register view data not provided')
+    }
+    const pupilRegisterRedisKey = `school:${schoolId}`
```
Comment (GeorgeChatzigiannis) [inline admin/services/pupil-register-caching.service.js]: Need thoughts on redis key name.
I am guessing it's not very descriptive 🤔

File: admin/services/pupil-register-caching.service.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+'use strict'
+
+const featureToggles = require('feature-toggles')
+
+const pupilRegisterV2Service = require('./pupil-register-v2.service')
+const pupilRegisterService = require('./pupil-register.service')
+const redisCacheService = require('./data-access/redis-cache.service')
+
+const pupilRegisterCachingService = {
```
Comment (GuyHarwood) [inline admin/services/pupil-register-caching.service.js]: i don't believe we need a dedicated pupil register cache, all the logic should be encapsulated in the pupil register service.

File: admin/services/pupil-register-v2.service.js [hunk 3]
```diff
@@ -11,6 +13,23 @@ const pupilRegisterV2Service = {
    * @return {Promise<*>}
    */
   getPupilRegister: async function (schoolId) {
+    if (!schoolId) {
+      throw new Error('School id not found in session')
+    }
+    const pupilRegisterRedisKey = `school:${schoolId}`
```
Comment (GuyHarwood) [inline admin/services/pupil-register-v2.service.js]: please add more specific key

File: admin/services/pupil-register-v2.service.js [hunk 4]
```diff
@@ -22,7 +41,10 @@ const pupilRegisterV2Service = {
         group: d.groupName
       }
     })
-    return pupilIdentificationFlagService.addIdentificationFlags(tableSorting.applySorting(pupilRegister, 'lastName'))
+    const pupilRegisterViewData = pupilIdentificationFlagService.addIdentificationFlags(tableSorting.applySorting(pupilRegister, 'lastName'))
+    const pupilRegisterRedisKey = `school:${schoolId}`
+    await redisCacheService.set(pupilRegisterRedisKey, pupilRegisterViewData)
```
Comment (GuyHarwood) [inline admin/services/pupil-register-v2.service.js]: you should check if the data exists in the cache, and return it.  if it doesnt exist, get from sql, add to cache and return

File: admin/controllers/pupil.js [hunk 5]
```diff
@@ -62,6 +63,7 @@ const postAddPupil = async (req, res, next) => {
   res.locals.pageTitle = 'Add pupil'
   try {
     const pupil = await pupilAddService.addPupil(req.body, req.user.schoolId)
+    await pupilRegisterCachingService.refreshPupilRegisterCache(req.user.schoolId)
```
[the remark this answers is not in the document]
Comment (GeorgeChatzigiannis) [reply admin/controllers/pupil.js]: @GuyHarwood this method and service is removed and replaced with redisservice.drop
When making a request to get the list of pupils in pupil register view the cache will be recreated.

File: admin/services/pupil-register-caching.service.js (hunk 2 above)
Comment (GeorgeChatzigiannis) [reply admin/services/pupil-register-caching.service.js]: @GuyHarwood yes, I removed it.

File: admin/services/pupil-register-v2.service.js (hunk 4 above)
Comment (GeorgeChatzigiannis) [reply admin/services/pupil-register-v2.service.js]: @GuyHarwood This is done by the `getPupilRegister` which calls `getPupilRegisterViewData` once it identifies that no data exist in the cache.
If `getPupilRegister` detects data in cache `getPupilRegisterViewData` won't be called at all.

File: admin/services/pupil-register-v2.service.js [hunk 6]
```diff
@@ -11,6 +13,24 @@ const pupilRegisterV2Service = {
    * @return {Promise<*>}
    */
   getPupilRegister: async function (schoolId) {
+    if (!schoolId) {
+      throw new Error('School id not found in session')
+    }
+    const pupilRegisterRedisKey = `pupilRegisterViewData:${schoolId}`
```
Comment (jon-shipley) [inline admin/services/pupil-register-v2.service.js]: I am slightly concerned that there isn't a buildKey() method or something.  If you GET the key here; how is it set? Duplicate code?

Comment (GuyHarwood) [reply admin/services/pupil-register-v2.service.js]: seems like a good opportunity to implement the `redisKeyService` we talked about.

`getPupilRegisterViewDataKey(schoolid: number): string`

File: admin/controllers/pupil.js [hunk 7]
```diff
@@ -62,6 +61,8 @@ const postAddPupil = async (req, res, next) => {
   res.locals.pageTitle = 'Add pupil'
   try {
     const pupil = await pupilAddService.addPupil(req.body, req.user.schoolId)
+    const pupilRegisterRedisKey = `pupilRegisterViewData:${req.user.schoolId}`
+    await redisCacheService.drop(pupilRegisterRedisKey)
```
Comment (GuyHarwood) [inline admin/controllers/pupil.js]: we should be invalidating cache in the service, not the controller

File: admin/controllers/pupil.js [hunk 8]
```diff
@@ -160,6 +161,8 @@ const postAddMultiplePupils = async (req, res, next) => {
     res.fileErrors = uploadResult.fileErrors
     return getAddMultiplePupils(req, res, next)
   } else {
+    const pupilRegisterRedisKey = `pupilRegisterViewData:${req.user.schoolId}`
+    await redisCacheService.drop(pupilRegisterRedisKey)
```
Comment (GuyHarwood) [inline admin/controllers/pupil.js]: we should be invalidating cache in the service, not the controller

File: admin/services/pupil-register-v2.service.js (hunk 6 above)
Comment (GeorgeChatzigiannis) [reply admin/services/pupil-register-v2.service.js]: added

File: admin/controllers/pupil.js (hunk 7 above)
Comment (GeorgeChatzigiannis) [reply admin/controllers/pupil.js]: @GuyHarwood updated

File: admin/services/group.service.js [hunk 9]
```diff
@@ -90,6 +93,8 @@ groupService.update = async (id, group, schoolId) => {
     // only update pupils if list has changed
     await groupDataService.sqlModifyGroupMembers(id, pupils)
   }
+  const pupilRegisterRedisKey = redisKeyService.getPupilRegisterViewDataKey(schoolId)
+  await redisCacheService.drop(pupilRegisterRedisKey)
```
Comment (jon-shipley) [inline admin/services/group.service.js]: This reads very well ⭐️

**Pull request merged**